pkg/services/sqlstore/datasource.go

```diff
@@ -67,8 +67,12 @@ func GetDataSourceByName(query *models.GetDataSourceByNameQuery) error {
 }
 
 func GetDataSources(query *models.GetDataSourcesQuery) error {
-	sess := x.Limit(5000, 0).Where("org_id=?", query.OrgId).Asc("name")
-
+	var sess *xorm.Session
+	if query.DataSourceLimit == nil {
+		sess = x.Where("org_id=?", query.OrgId).Asc("name")
```

Comment:
I wonder if this would be too costly if there is a huge amount of data sources.
Off course the table has an [index](https://github.com/grafana/grafana/blob/aebe8985c57f7b24b7245066d85e1bf3b5b58313/pkg/services/sqlstore/migrations/datasource_mig.go#L67) on the `org_id` column but anyway since the previous behaviour was to fetch only the first 5000
I would suggest to fetch all the datasources in batches of 5000 (using a loop).
@aknuds1 WDYT?

Reply:
@papagian To tell you the truth I'm unsure. Why is it less costly to fetch in batches? Does it not impose the same load on the whole (we still populate a full array with all the results)? @kylebrandt Do you have any insight to share in this regard?

Reply:
@papagian BTW thinking more about this, doesn't this also cause a change in behaviour? I.e., if the caller hasn't set DataSourceLimit, the query will be unlimited. But those callers that uses the configuration will always have set a non-nil `DataSourceLimit`, so those cases will never be unlimited.

As such, I think that we need to retain the old behaviour of 5000 as default limit.

Reply:
yes this is called by the api and the alerting service. For alerting I think that there should no limit applied for the reasons I have described [here](https://github.com/grafana/grafana/issues/29262#issuecomment-732972406).

Reply:
@papagian If we shouldn't retain the old limit, we should use the value of 0 to represent no limit. Then we would obtain homogeneous behaviour if the user configures 0 for the limit also, i.e. that it's unlimited, rather than actually limited at 0, which would happen now.

Reply:
It is used by the json package for distinguishing between missing/null and zero values and I had got the impression that it was generally used but it seems that it's not that common.
Anyway given that 0 value is no limit I agree that it is better not to use a pointer and in that case we don't need even this conditional.

Reply:
I've updated my comment above in order to align with [this](https://github.com/grafana/grafana/pull/29358/files#r530834486) comment.

Reply:
> I would suggest to fetch all the datasources in batches of 5000 (using a loop). was to fetch only the first 5000

I think it's fine to fetch all in this case since 5k will remain the default limit. Anyone who increases that limits is operating outside our recommendation and doing something weird. Paging will eventually be the long-term solution while this is just a temporary workaround IMO. Perhaps we should add a note about this in the config file/settings 

> yes, this is called by the api and the alerting service. For alerting I think that there should no limit applied for the reasons I have described here.

Seems reasonable. Perhaps we could cache the query instead of running it once for every alert rule if dsname is missing? The best solution would ofc be to migrate all those alert rules that don't have a datasource name. But I have a feeling an alert rule created the last three years have data source name or id configured :)

> I'm pretty certain that the Go idiom is to generally use zero values instead of nil pointers to represent defaults, but I'm unable to google it up.

https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful

In general, I dont think we have to think too much about performance optimization for instances more than 5000k datasources.